hi-ml-cpath/src/health_cpath/utils/callbacks.py

```diff
@@ -200,67 +196,92 @@ def gather_loss_cache(self, rank: int) -> None:
-
-        return np.array(slides).T, np.array(slides_loss).T
+        return self._select_values_for_epoch(ResultsKey.SLIDE_ID, epoch, high, num_slides)
 
-    def select_loss_for_slides_of_epoch(self, epoch: int, high: Optional[bool] = None) -> LossDictType:
-        """Selects the slides with the highest/lowest loss values for a given epoch and returns the loss values for each
-        slide across all epochs.
+    def select_slides_losses_for_epoch(
+        self, epoch: int, high: Optional[bool] = None, num_slides: Optional[int] = None
+    ) -> np.ndarray:
+        """Selects loss values of slides in ascending/descending order of loss at a given epoch
 
         :param epoch: The epoch to select the slides from.
         :param high: If True, selects the slides with the highest loss values, else selects the slides with the lowest
             loss values. If None, selects all slides.
-        :return: A dictionary containing the loss values for each slide across all epochs.
+        :param num_slides: The number of slides to select. If None, selects all slides.
         """
-        loss_cache = pd.read_csv(self.cache_folder / LOSS_VALUES_FILENAME.format(self.zfill_epoch(epoch)))
+        return self._select_values_for_epoch(ResultsKey.LOSS, epoch, high, num_slides)
 
-        if high:
-            slides = loss_cache[ResultsKey.SLIDE_ID][: self.num_slides_heatmap]
-        elif high is False:
-            slides = loss_cache[ResultsKey.SLIDE_ID][-self.num_slides_heatmap:]
-        else:
-            slides = loss_cache[ResultsKey.SLIDE_ID]
+    def select_all_losses_for_selected_slides(self, slides: np.ndarray) -> LossDictType:
+        """Selects the loss values for a given set of slides"""
 
         slides_loss_values: LossDictType = {slide_id: [] for slide_id in slides}
         for epoch in self.epochs_range:
-            loss_cache = pd.read_csv(self.cache_folder / LOSS_VALUES_FILENAME.format(self.zfill_epoch(epoch)))
-            loss_cache.set_index(ResultsKey.SLIDE_ID, inplace=True)
+            loss_cache = self.read_loss_cache(epoch, idx_col=ResultsKey.SLIDE_ID)
             slides_loss_values = {
                 slide_id: slides_loss_values[slide_id] + [loss_cache.loc[slide_id, ResultsKey.LOSS]]
                 for slide_id in slides
             }
         return slides_loss_values
 
+    def select_slides_and_losses_across_epochs(self, high: bool = True) -> Tuple[np.ndarray, np.ndarray]:
+        """Selects the slides with the highest/lowest loss values across epochs
+
+        :param high: If True, selects the slides with the highest loss values, else selects the slides with the lowest
+        loss values.
+        """
+        slides = []
+        slides_loss = []
+        for epoch in self.epochs_range:
+            epoch_slides = self.select_slides_for_epoch(epoch, high, self.num_slides_scatter)
+            epoch_slides_loss = self.select_slides_losses_for_epoch(epoch, high, self.num_slides_scatter)
```

Comment:
This will call `read_loss_cache()` twice for each `epoch` (once for slide IDs and once for losses), no? Do you expect this to have any impact on performance?